plugins/processors/gpuattributes/processor.go

```diff
@@ -0,0 +1,194 @@
+)
+
+// schemas at each resource level
+// - Container Schema
+//   - ClusterName
+//   - ClusterName, Namespace, PodName, ContainerName
+//   - ClusterName, Namespace, PodName, FullPodName, ContainerName
+//   - ClusterName, Namespace, PodName, FullPodName, ContainerName, GpuDevice
+//
+// - Pod
+//   - ClusterName
+//   - ClusterName, Namespace
+//   - ClusterName, Namespace, Service
+//   - ClusterName, Namespace, PodName
+//   - ClusterName, Namespace, PodName, FullPodName
+//   - ClusterName, Namespace, PodName, FullPodName, GpuDevice
+//
+// - Node
+//   - ClusterName
+//   - ClusterName, InstanceIdKey, NodeName
+//   - ClusterName, InstanceIdKey, NodeName, GpuDevice
+
+var commonLabels = []string{
+	containerinsightscommon.ClusterNameKey,
+	containerinsightscommon.InstanceIdKey,
+	containerinsightscommon.GpuDeviceKey,
+	containerinsightscommon.MetricType,
+	containerinsightscommon.NodeNameKey,
+	containerinsightscommon.VersionKey,
+	containerinsightscommon.SourcesKey,
+	containerinsightscommon.Timestamp,
+}
+
+var podAndContainerLabels = []string{
+	containerinsightscommon.K8sNamespace,
+	containerinsightscommon.FullPodNameKey,
+	containerinsightscommon.PodNameKey,
+	containerinsightscommon.TypeService,
+	containerinsightscommon.GpuUniqueId,
+}
+
+var containerK8sBlobLabels = []string{
+	"container_name",
+	"containerd",
+}
+var podK8sBlobLabels = []string{
+	"host",
+	"labels",
+	"pod_id",
+	"pod_name",
+	"pod_owners",
+	"namespace",
+}
+
+type gpuAttributesProcessor struct {
+	*Config
+	logger *zap.Logger
+}
+
+func newGpuAttributesProcessor(config *Config, logger *zap.Logger) *gpuAttributesProcessor {
+	d := &gpuAttributesProcessor{
+		Config: config,
+		logger: logger,
+	}
+	return d
+}
+
+func (d *gpuAttributesProcessor) processMetrics(_ context.Context, md pmetric.Metrics) (pmetric.Metrics, error) {
+	rms := md.ResourceMetrics()
+	for i := 0; i < rms.Len(); i++ {
+		rs := rms.At(i)
+		ilms := rs.ScopeMetrics()
+		for j := 0; j < ilms.Len(); j++ {
+			ils := ilms.At(j)
+			metrics := ils.Metrics()
+			for k := 0; k < metrics.Len(); k++ {
+				m := metrics.At(k)
+				d.processMetricAttributes(m)
+			}
+		}
+	}
+	return md, nil
+}
+
+func (d *gpuAttributesProcessor) processMetricAttributes(m pmetric.Metric) {
+	// only decorate GPU metrics
+	if !strings.Contains(m.Name(), gpuMetricIdentifier) {
+		return
+	}
+
+	var labels []string
+	labels = append(labels, commonLabels...)
+	k8sBlobLabels := []string{containerinsightscommon.HostKey}
+	if strings.HasPrefix(m.Name(), gpuContainerMetricPrefix) {
+		labels = append(labels, podAndContainerLabels...)
+		labels = append(labels, containerinsightscommon.ContainerNamekey)
+		k8sBlobLabels = append(k8sBlobLabels, containerK8sBlobLabels...)
+		k8sBlobLabels = append(k8sBlobLabels, podK8sBlobLabels...)
+	} else if strings.HasPrefix(m.Name(), gpuPodMetricPrefix) {
+		labels = append(labels, podAndContainerLabels...)
+		k8sBlobLabels = append(k8sBlobLabels, podK8sBlobLabels...)
+	}
```

Comment:
You should pre-compute `labelFilter` for the different metric types instead of redoing it every time the hot path of this processor.

Reply:
yeah, converted them vars